[Member] Member Application 계층 구현

vwx12345 opened the pull request:
### 💻 작업 내용
- Member의 Command와 Query를 분리하여 구현했습니다.

### ✨ 리뷰 포인트
- 

### 📝 메모
- 간단한 코드라 테스트는 따로 작성하지 않았습니다.
- MemberRepository의 구현체가 아직 없습니다.

### 🎯 관련 이슈
closed #69

vwx12345 committed:
feat(member): MemberCommandService 구현

vwx12345 committed:
feat(member): Member 도메인 이벤트 객체 구현

vwx12345 committed:
feat(member): Member Query application 구현

vwx12345 committed:
feat(member): MemberNotFound Error 정의

vwx12345 committed:
docs(member): 주석 작성

codingbaraGo commented on src/main/java/com/simpleboard/board/member/application/command/MemberCommandService.java:
```diff
@@ -0,0 +1,52 @@
+    eventPublisher.publishEvent(new MemberWithdrawnEvent(memberId));
+  }
+
+  public void updateNickname(Long memberId, String newNickname) {
+    Member member =
+        memberRepository
+            .findById(memberId)
+            .orElseThrow(MemberNotFoundException::new);
+
+    nicknameDuplicationChecker.checkDuplicated(newNickname);
+    member.updateNickname(newNickname);
```
Board B.C의 도메인 파트에도 비슷한 코드가 있던것으로 기억합니다.
Board B.C에서는 aggregate에 대한 action(create 메서드) 속에서 duplication check를 수행했던 것으로 기억하는데 여기서는 application 계층에서 check를 진행 후 action(updateNickname)을 진행하네요.

제가 생각하기에는 둘의 duplication check 구조가 유사한듯 한데 어떤 부분에서 이러한 차이가 발생했을까요?

만약 action의 완결성을 추구하셨다면 member B.C에서도 validation을 aggregate 내부에서 하는편이 더 좋고, 기능의 독립성과 흐름의 명확성을 추구하셨다면 Board B.C에서도 지금처럼 application 계층에서 checker를 사용 후 action을 호출하는게 더 좋지 않을까 해서요.
혹은 다른 기준이 있으셨나요?

vwx12345 replied:
맞습니다. 저는 Board의 2번 방식이 DDD에 더 맞다고 봅니다. 규칙은 도메인에서 강제하고, 구현은 포트/어댑터로 자유롭게 교체할 수 있어서요. 또한 프레임워크 애노테이션 없이 순수 도메인으로 유지할 수 있다는 점도 장점입니다. Member 쪽 코드는 기존 구현을 재사용하려다 보니 애플리케이션에서 체크하는 형태가 되었는데, Board와 동일한 패턴으로 리팩터링하는 게 맞다고 생각합니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/member/application/query/MemberQueryService.java:
```diff
@@ -0,0 +1,28 @@
+  public MemberQueryService(MemberQueryRepository memberQueryRepository) {
+    this.memberQueryRepository = memberQueryRepository;
+  }
+
+  public Boolean existsByNickname(String nickname) {
+    return memberQueryRepository.existsByNickname(nickname);
+  }
```
이 메서드는 어떤 Usecase에서 사용하는건가요?
회원가입쪽 닉네임 중복체크와 닉네임 변경시 사용하는건가요?

codingbaraGo replied:
DDD의 Application 계층은 도메인 모델을 직접 노출하지 않고, 외부에서 요청된 유스케이스 단위의 작업을 조합·조율하는 계층입니다.
Query, Command 양쪽 다 말이죠.
주석으로 어떤 usecase를 처리하는 메서드인지 간단하게라도 적어주면 좋을것같아요.

vwx12345 replied:
> 이 메서드는 어떤 Usecase에서 사용하는건가요? 회원가입쪽 닉네임 중복체크와 닉네임 변경시 사용하는건가요?

현재 생각나는 케이스는 Board 에서 권한 이전 시에 닉네임 기반으로 실제 사용자가 존재하는지 확인용입니다. 회원가입 쪽 닉네임 중복체크와 닉네임 변경시에 사용되지는 않습니다.

vwx12345 replied:
> DDD의 Application 계층은 도메인 모델을 직접 노출하지 않고, 외부에서 요청된 유스케이스 단위의 작업을 조합·조율하는 계층입니다. Query, Command 양쪽 다 말이죠. 주석으로 어떤 usecase를 처리하는 메서드인지 간단하게라도 적어주면 좋을것같아요.

저는 이 정도까지 주석을 달 필요는 없다고 생각합니다. existsByNickname이라는 메서드명만으로도 닉네임 존재 여부를 확인한다는 의미가 충분히 전달됩니다. 여기에 “어떤 유스케이스에서 사용되는지”까지 적게 되면, 추후 이 메서드가 다른 유스케이스에서도 활용될 때 주석을 계속 수정해야 하고 이는 오히려 SRP를 해칠 수 있다고 봅니다.

codingbaraGo replied:
> > DDD의 Application 계층은 도메인 모델을 직접 노출하지 않고, 외부에서 요청된 유스케이스 단위의 작업을 조합·조율하는 계층입니다. Query, Command 양쪽 다 말이죠. 주석으로 어떤 usecase를 처리하는 메서드인지 간단하게라도 적어주면 좋을것같아요.
> 
> 저는 이 정도까지 주석을 달 필요는 없다고 생각합니다. existsByNickname이라는 메서드명만으로도 닉네임 존재 여부를 확인한다는 의미가 충분히 전달됩니다. 여기에 “어떤 유스케이스에서 사용되는지”까지 적게 되면, 추후 이 메서드가 다른 유스케이스에서도 활용될 때 주석을 계속 수정해야 하고 이는 오히려 SRP를 해칠 수 있다고 봅니다.

Application Service는 본질적으로 유스케이스 단위의 작업을 조합·조율하는 계층입니다. 따라서 메서드 자체가 특정 유스케이스를 처리한다는 점이 중요합니다.

existsByNickname이라는 이름만으로도 “닉네임 존재 여부 확인” 의미는 충분히 드러납니다. 다만 이 메서드가 어떤 유스케이스를 위해 존재하는지는 이름만으로 파악하기 어렵습니다. Repository나 Domain Service였다면 이러한 부분이 납득 가능했겠지만, Application Service는 유스케이스 중심으로 설계되므로 상황이 다릅니다.

예를 들어 같은 닉네임 중복 검사라도 유스케이스에 따라 단순 조회만 하거나, lock 처리까지 할 수도 있고, 클라이언트에 boolean을 반환할지, 혹은 Exception을 이용해 반환할지 등 구현 방식이 달라질 수 있습니다. 결국 응용 서비스 메서드가 어느 유스케이스를 책임지는지가 설계 판단 기준이 됩니다.

특수한 케이스(internal/external API에서 공통 사용 등) 외에는 보통 한 응용 메서드가 하나의 유스케이스를 담당하도록 두는 게 일반적입니다. 같은 기능을 수행하는 유스케이스들을 분리해서 둘 이유가 없기 때문이죠.

이러한 이유에서 저는 주석으로 USECASE를 달아 놓는편이 좋다고 생각합니다.

vwx12345 replied:
> > > DDD의 Application 계층은 도메인 모델을 직접 노출하지 않고, 외부에서 요청된 유스케이스 단위의 작업을 조합·조율하는 계층입니다. Query, Command 양쪽 다 말이죠. 주석으로 어떤 usecase를 처리하는 메서드인지 간단하게라도 적어주면 좋을것같아요.
> > 
> > 
> > 저는 이 정도까지 주석을 달 필요는 없다고 생각합니다. existsByNickname이라는 메서드명만으로도 닉네임 존재 여부를 확인한다는 의미가 충분히 전달됩니다. 여기에 “어떤 유스케이스에서 사용되는지”까지 적게 되면, 추후 이 메서드가 다른 유스케이스에서도 활용될 때 주석을 계속 수정해야 하고 이는 오히려 SRP를 해칠 수 있다고 봅니다.
> 
> Application Service는 본질적으로 유스케이스 단위의 작업을 조합·조율하는 계층입니다. 따라서 메서드 자체가 특정 유스케이스를 처리한다는 점이 중요합니다.
> 
> existsByNickname이라는 이름만으로도 “닉네임 존재 여부 확인” 의미는 충분히 드러납니다. 다만 이 메서드가 어떤 유스케이스를 위해 존재하는지는 이름만으로 파악하기 어렵습니다. Repository나 Domain Service였다면 이러한 부분이 납득 가능했겠지만, Application Service는 유스케이스 중심으로 설계되므로 상황이 다릅니다.
> 
> 예를 들어 같은 닉네임 중복 검사라도 유스케이스에 따라 단순 조회만 하거나, lock 처리까지 할 수도 있고, 클라이언트에 boolean을 반환할지, 혹은 Exception을 이용해 반환할지 등 구현 방식이 달라질 수 있습니다. 결국 응용 서비스 메서드가 어느 유스케이스를 책임지는지가 설계 판단 기준이 됩니다.
> 
> 특수한 케이스(internal/external API에서 공통 사용 등) 외에는 보통 한 응용 메서드가 하나의 유스케이스를 담당하도록 두는 게 일반적입니다. 같은 기능을 수행하는 유스케이스들을 분리해서 둘 이유가 없기 때문이죠.
> 
> 이러한 이유에서 저는 주석으로 USECASE를 달아 놓는편이 좋다고 생각합니다.

그렇다면 existsByNickname()이라는 메서드명을 isUserExistsByNickname() 로 바꾸는건 어떻게 생각하시나요?
주석으로 "이 메서드가 어떤 유스케이스를 담당한다"라고 달아둘 수도 있겠지만, 개인적으로는 주석보다 네이밍 자체가 유스케이스를 설명하는 편을 선호합니다. 또한 Application Service는 말씀하신 것처럼 유스케이스 단위의 작업을 책임지는 계층이기 때문에, 메서드명에서부터 그 의도가 드러나는 것이 가장 명확하다고 봅니다.

codingbaraGo replied:
> > > > DDD의 Application 계층은 도메인 모델을 직접 노출하지 않고, 외부에서 요청된 유스케이스 단위의 작업을 조합·조율하는 계층입니다. Query, Command 양쪽 다 말이죠. 주석으로 어떤 usecase를 처리하는 메서드인지 간단하게라도 적어주면 좋을것같아요.
> > > 
> > > 
> > > 저는 이 정도까지 주석을 달 필요는 없다고 생각합니다. existsByNickname이라는 메서드명만으로도 닉네임 존재 여부를 확인한다는 의미가 충분히 전달됩니다. 여기에 “어떤 유스케이스에서 사용되는지”까지 적게 되면, 추후 이 메서드가 다른 유스케이스에서도 활용될 때 주석을 계속 수정해야 하고 이는 오히려 SRP를 해칠 수 있다고 봅니다.
> > 
> > 
> > Application Service는 본질적으로 유스케이스 단위의 작업을 조합·조율하는 계층입니다. 따라서 메서드 자체가 특정 유스케이스를 처리한다는 점이 중요합니다.
> > existsByNickname이라는 이름만으로도 “닉네임 존재 여부 확인” 의미는 충분히 드러납니다. 다만 이 메서드가 어떤 유스케이스를 위해 존재하는지는 이름만으로 파악하기 어렵습니다. Repository나 Domain Service였다면 이러한 부분이 납득 가능했겠지만, Application Service는 유스케이스 중심으로 설계되므로 상황이 다릅니다.
> > 예를 들어 같은 닉네임 중복 검사라도 유스케이스에 따라 단순 조회만 하거나, lock 처리까지 할 수도 있고, 클라이언트에 boolean을 반환할지, 혹은 Exception을 이용해 반환할지 등 구현 방식이 달라질 수 있습니다. 결국 응용 서비스 메서드가 어느 유스케이스를 책임지는지가 설계 판단 기준이 됩니다.
> > 특수한 케이스(internal/external API에서 공통 사용 등) 외에는 보통 한 응용 메서드가 하나의 유스케이스를 담당하도록 두는 게 일반적입니다. 같은 기능을 수행하는 유스케이스들을 분리해서 둘 이유가 없기 때문이죠.
> > 이러한 이유에서 저는 주석으로 USECASE를 달아 놓는편이 좋다고 생각합니다.
> 
> 그렇다면 existsByNickname()이라는 메서드명을 isUserExistsByNickname() 로 바꾸는건 어떻게 생각하시나요? 주석으로 "이 메서드가 어떤 유스케이스를 담당한다"라고 달아둘 수도 있겠지만, 개인적으로는 주석보다 네이밍 자체가 유스케이스를 설명하는 편을 선호합니다. 또한 Application Service는 말씀하신 것처럼 유스케이스 단위의 작업을 책임지는 계층이기 때문에, 메서드명에서부터 그 의도가 드러나는 것이 가장 명확하다고 봅니다.

좋습니다. 메서드 명을 바꾸니 좀 더 이해가 명확하게 되는 것 같네요.

codingbaraGo commented on src/main/java/com/simpleboard/board/member/application/query/repository/MemberQueryRepository.java:
```diff
@@ -0,0 +1,10 @@
+public interface MemberQueryRepository {
+  boolean existsByNickname(String nickname);
+
+  Optional<Member> findById(Long memberId);
```
Query repository에 aggregate root 엔티티를 반환하는 메서드는 필요없지 않을까요?
도메인 엔티티를 그대로 조회할거라면 그냥 JPA를 쓰는편이 훨씬 편할것같아요.

vwx12345 replied:
저도 Query Repository에서 애그리게이트 루트를 그대로 반환하는 건 적절하지 않다고 생각합니다. 해당 부분은 다시 수정하여 pr 올리겠습니다.

codingbaraGo reviewed (commented):
커밋이 아주 명확해서 정말 좋습니다.
고생하셨습니다.

vwx12345 committed:
refactor(member): Member domain 을 QueryRepository 에서 직접 조회하는 문제 해결

- MemberFullView 를 통해서 Member 정보를 가져오는 방식으로 변경

vwx12345 committed:
refactor(member): NicknamePolicy 로 변경

- 기존의 NicknameDuplicationChecker 에서 NicknamePolicy 를 통해 Nickname을 검증하는 방식으로 변경

vwx12345 committed:
refactor(member): Nickname hashCode 와 equals 직접 구현

- 이름 기반으로 검증을 위해 hashCode와 equals 직접 구현

vwx12345 committed:
test(member): NicknamePolicy 로 테스트 코드 수정

vwx12345 committed:
refactor(member): BirthYear Integer 기반으로 변경

- 기존 년도 값을 String 에서 Integer type 으로 변경

vwx12345 committed:
refactor(member): findById method 명 변경

- findById 보다는 findFullById 로 도메인이 뷰 모델임을 드러나게 변경

vwx12345 committed:
directory(member): Event 위치 변경

- application 에서 command 에서 사용되는 event 여서 command 하위로 변경

vwx12345 committed:
feat(member): member 등록 추가

- registerMember 를 통해서 member 등록 추가

vwx12345 committed:
refactor(member): existsByNickname 메서드 이름 변경

- 메서드의 의미를 더 명확히 하기 위해 이름을 변경
- 기존: existsByNickname
- 변경: isUserExistsByNickname
- '닉네임'보다는 '유저 존재 여부'에 초점을 두도록 개선

vwx12345 committed:
feat(member): MemberRegisterCommand 추가

codingbaraGo reviewed (approved):
고생하셨습니다.
